src/mlpack/methods/fastmks/fastmks_impl.hpp

```diff
@@ -174,21 +212,54 @@ template<typename KernelType,
                   typename TreeMatType> class TreeType>
 void FastMKS<KernelType, MatType, TreeType>::Train(Tree* tree)
 {
+  train(*tree);
+}
+
+template<typename KernelType,
+         typename MatType,
+         template<typename TreeMetricType,
+                  typename TreeStatType,
+                  typename TreeMatType> class TreeType>
+void FastMKS<KernelType, MatType, TreeType>::Train(Tree& tree)
+{
   if (naive)
     throw std::invalid_argument("cannot call FastMKS::Train() with a tree when "
         "in naive search mode");
 
   if (setOwner)
-    delete this->referenceSet;
-
-  this->referenceSet = &tree->Dataset();
-  this->metric = metric::IPMetric<KernelType>(tree->Metric().Kernel());
-  this->setOwner = false;
+    delete referenceSet;
 
   if (treeOwner && referenceTree)
     delete referenceTree;
+    
+  referenceTree = new Tree(tree);    
+  referenceSet = &tree->Dataset();
```

Comment:
I bet the problem is in this line.  Shouldn't this be `referenceSet = &referenceTree->Dataset()`?  Otherwise you are setting the pointer to the dataset of the tree that the user passed in, which isn't guaranteed to be a valid pointer outside the scope of this function.